### PR TITLE
Feat/lnist/pooled forwarder

### PR DESCRIPTION
**Pull request by lnist**

### CODE REVIEW

File: src/strategies/offer_forwarder/PooledForwarder.sol [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    AbstractRouter router_ = router();
+    router_.bind(address(this));
+    setAdmin(deployer);
+    router_.setAdmin(deployer); // consider if admin should be this contract?
+  }
+
+  // Increases balance of token for owner.
+  function increaseBalance(IERC20 token, address owner, uint amount) private returns (uint) {
+    uint newBalance = ownerBalance[token][owner] + amount;
+    ownerBalance[token][owner] = newBalance;
```
Comment (jkrivine) [inline src/strategies/offer_forwarder/PooledForwarder.sol:106]: or just `ownerBalance[token][owner] += amount`

Comment (jkrivine) [reply src/strategies/offer_forwarder/PooledForwarder.sol:106]: Also you it is good practice to log balance change using specific event

File: src/strategies/offer_forwarder/PooledForwarder.sol [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    require(currentBalance >= amount, "PooledForwarder/decreaseBalance/amountMoreThanBalance");
+    unchecked {
+      uint newBalance = currentBalance - amount;
+      ownerBalance[token][owner] = newBalance;
+      return newBalance;
+    }
+  }
+
+  // TODO will this be inlined?
+  function getBalance(IERC20 token, address owner) public view returns (uint) {
```
Comment (jkrivine) [inline src/strategies/offer_forwarder/PooledForwarder.sol:122]: this amounts to making ownerBalance `public`

File: src/strategies/offer_forwarder/PooledForwarder.sol [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+  function increaseBalance(IERC20 token, address owner, uint amount) private returns (uint) {
+    uint newBalance = ownerBalance[token][owner] + amount;
+    ownerBalance[token][owner] = newBalance;
+    return newBalance;
+  }
+
+  // Decrease balance of token for owner.
+  function decreaseBalance(IERC20 token, address owner, uint amount) private returns (uint) {
+    uint currentBalance = ownerBalance[token][owner];
+    require(currentBalance >= amount, "PooledForwarder/decreaseBalance/amountMoreThanBalance");
```
Comment (jkrivine) [inline src/strategies/offer_forwarder/PooledForwarder.sol]: I am afraid the revert reason is too long and will be truncated by Mangrove when cast to 32 bytes

File: src/strategies/offer_forwarder/PooledForwarder.sol [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+
+  function __posthookFallback__(MgvLib.SingleOrder calldata order, MgvLib.OrderResult calldata)
+    internal
+    override
+    returns (bytes32)
+  {
+    return pushAllToAave(order);
+  }
+
+  function pushAllToAave(MgvLib.SingleOrder calldata order) internal returns (bytes32) {
```
Comment (peterMangrove) [inline src/strategies/offer_forwarder/PooledForwarder.sol]: We are not returning a byte32? Should we not return some kind of info?

File: src/strategies/offer_forwarder/PooledForwarder.sol [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+    return true;
+  }
+
+  function __posthookSuccess__(MgvLib.SingleOrder calldata order, bytes32 makerData)
+    internal
+    override
+    returns (bytes32)
+  {
+    // reposts residual if any (conservative hook)
+    bytes32 repost_status = super.__posthookSuccess__(order, makerData);
```
Comment (peterMangrove) [inline src/strategies/offer_forwarder/PooledForwarder.sol]: Why define a variable, if we are just ignoring it? Why create it in the first place?

**Review by peterMangrove — APPROVED** [review]

**Pull request closed without merging**